src/modules/audio/audio.c

```diff
@@ -31,90 +42,192 @@ struct Source {
 
 static struct {
   bool initialized;
   ma_context context;
-  AudioConfig config[AUDIO_TYPE_COUNT];
+  AudioDeviceConfig deviceConfig[AUDIO_TYPE_COUNT];
   ma_device devices[AUDIO_TYPE_COUNT];
   ma_mutex playbackLock;
   Source* sources;
   ma_pcm_rb captureRingbuffer;
   arr_t(ma_data_converter*) converters;
   Spatializer* spatializer;
+  bool fixedBuffer;
+  uint32_t bufferSize;
+  float *scratchBuffer1, *scratchBuffer2; // Used internally by mix(). Contains bufferSize stereo frames.
+  float *persistBuffer;  // If fixedBuffer, preserves excess audio between frames.
+  float *persistBufferContent; // Pointer into persistBuffer
+  uint32_t persistBufferRemaining; // In fixedBuffer mode, how much of the previous frame's mixBuffer was consumed?
+
+#ifdef LOVR_DEBUG_AUDIOTAP
+  bool audiotapWriting;
+#endif
 } state;
 
 // Device callbacks
 
-static bool mix(Source* source, float* output, uint32_t count) {
-  float raw[2048];
-  float aux[2048];
-  float mix[4096];
-
-  // TODO
-  // frameLimitIn =
-  // frameLimitOut =
-
-  while (count > 0) {
-    uint32_t chunk = MIN(sizeof(raw) / SampleFormatBytesPerFrame(source->sound->channels, source->sound->format),
-        ma_data_converter_get_required_input_frame_count(source->converter, count));
-        // ^^^ Note need to min `count` with 'capacity of aux buffer' and 'capacity of mix buffer'
-        // could skip min-ing with one of the buffers if you can guarantee that one is bigger/equal to the other (you can because their formats are known)
-    ma_uint64 framesIn = source->sound->read(source->sound, source->offset, chunk, raw);
-    ma_uint64 framesOut = sizeof(aux) / (sizeof(float) * outputChannelCountForSource(source));
-
-    ma_data_converter_process_pcm_frames(source->converter, raw, &framesIn, aux, &framesOut);
+// Return value is number of stereo frames in buffer
+// Note: output is always equal to scratchBuffer1. This saves a little memory but is ugly.
+//       count is always less than or equal to the size of scratchBuffer1
+static int generateSource(Source* source, float* output, uint32_t count) {
+  char *raw; float *aux; // Scratch buffers: Raw generated audio from source; converted to float by converter
+  if (source->spatial) { // In spatial mode, raw and aux are mono and only output is stereo
+    raw = (char *)state.scratchBuffer1;
+    aux = state.scratchBuffer2;
+  } else {               // Otherwise, the data converter will produce stereo and aux=output is stereo
+    raw = (char *)state.scratchBuffer2;
+    aux = output;
+  }
 
-    if (source->spatial) {
-      state.spatializer->apply(source, source->transform, aux, mix, framesOut);
+  bool sourceFinished = false;
+  ma_uint64 framesIn = 0;
+  while (framesIn < count) { // Read from source until raw buffer filled
+    ma_uint64 framesRequested = count - framesIn;
+    // FIXME: Buffer size math will break (crash) if channels > 1
```

Comment:
Despite the comment this will work for channels <= 2, however this is still a problem. We could fix this by moving the ma_data_converter_process_pcm_frames call inside this loop. Is it possible to load a channels > 2 source or is there a check stopping this earlier?

Reply:
Currently only mono and stereo SoundData/Sources are supported.  However, I would like to support ambisonic sound formats soon, which can have many channels (I think 4 is common).  This may or may not cause problems, because I'm not sure if ambisonics will fall under the same codepaths in the mixer or if they'll need special handling.